marker_api/src/ast/common/span.rs

```diff
@@ -2,133 +2,469 @@ use std::marker::PhantomData;
+}
+
+/// Information about a specific expansion.
+///
+/// [`Span`]s in Rust are structured in layers. The root layer is the source code
+/// written in a source file. Macros can be expanded to AST nodes with [`Span`]s.
+/// These are then added as a new layer, on top of the root. This struct provides
+/// the information about one expansion layer.
+///
+/// ### Simple Macro Rules
+///
+/// ```
+/// macro_rules! ex1 {
+///     () => {
+///         1 + 1
+///     };
+/// }
+///
+/// ex1!();
+/// ```
+///
+/// In this example `ex1!()` expands into the expression `1 + 1`. The [`Span`]
+/// of the binary expression and numbers will all be from an expansion. Snipping the
+/// [`Span`] of the binary expression would return `1 + 1` from inside the macro rules.
+///
+/// ### Macro Rules with Parameters
+///
+/// ```
+/// macro_rules! ex2 {
+///     ($a:literal, $b:literal) => {
+///         $a + $b
+///     };
+/// }
+///
+/// ex2!(1, 2);
+/// ```
+///
+/// In this example `ex2!(1, 2)` expands into the expression `1 + 2`. The [`Span`]
+/// of the binary expression is marked to come from the expansion of a macro.
+/// The `1` and `2` literals are marked as coming from the root layer, since
+/// they're actually written in the source file.
+///
+/// ### Macros Invoking Macros
+///
+/// ```
+/// macro_rules! ex3a {
+///     ($a:literal, $b:literal) => {
+///         $a + $b
+///     };
+/// }
+/// macro_rules! ex3b {
+///     ($a:literal) => {
+///         ex3a!($a, 3)
+///     };
+/// }
+///
+/// ex3b!(2);
+/// ```
+///
+/// In this example `ex3b!(2)` expands to `ex3a(2, 3)` which in turn expands to
+/// `2 + 3`. This expansion has three layers, first the root, which contains the
+/// `ex3b!(2)` call. The next layer is the `ex3a!(2, 3)` call. The binary expression
+/// comes from the third layer, the number 3 from the second, and the number 2 from
+/// the root layer, since this one was actually written by the user.
+///
+/// ### Macros Creating Macros
+///
+/// ```
+/// macro_rules! ex4a {
+///     () => {
+///         macro_rules! ex4b {
+///             () => {
+///                 4 + 4
+///             };
+///         }
+///     };
+/// }
+///
+/// ex4a!();
+/// ex4b!();
+/// ```
+///
+/// This example expands `ex4a` into a new `ex4b` macro, which in turn expands
+/// into a `4 + 4` expression. The [`Span`] of the binary expression has three
+/// layers. First the root layer calling the `ex4b` macro, this would then
+/// expand to the `4 + 4` expression, which actually comes from the `ex4a` expansion.
+///
+/// ### Proc Macros
+///
+/// Proc macros and some built-in macros are different from `macro_rules!` macros as
+/// they are opaque to the driver. It's just known that some tokens are provided as an
+/// input and somehow expanded. The [`Span`]s of the expanded tokens are marked as
+/// coming from an expansion by default. However, macro crates can sometimes override
+/// this with some trickery. (Please use this forbidden knowledge carefully.)
```

Comment:
This is referring to `quote_spanned`, which can be a gigantic headache for linting. Clippy had several FPs and at times, just had to snip the source code to see if it matches with the AST. For linting reasons, I would either advise against using it at all, or making sure to always add `#[allow(clippy::all, clippy::pedantic)]`. This works with Clippy being the only linter, but it could again become a real problem for Marker

Reply:
I see, so it's not easily possible to detect if the code comes from the proc macro expansion, and therefore the author of the proc macro should add an `#[allow(...)]` to work around that. That's a pity, we should definitely not rely on people adding `allow(...)` in their macros. Checking the source snippet for the presence of a macro may be better

Reply:
Jup, there is some discussion to maybe add more support in rustc, we'll see

Reply:
> Checking the source snippet for the presence of a macro may be better

This is what Clippy currently does, but it sadly takes a significant amount of time.